[rustdoc] Add `--output-format=doctest` command-line flag

GuillaumeGomez opened the pull request:
Part of https://github.com/rust-lang/rust/issues/134529.

It was discussed with the Rust-for-Linux project recently that they needed a way to extract doctests so they can modify them and then run them more easily (look for "a way to extract doctests" [here](https://github.com/Rust-for-Linux/linux/issues/2)).

For now, I output most of `ScrapedDoctest` fields in JSON format with `serde_json`. So it outputs the following information:

 * filename
 * line
 * langstr
 * text

cc @ojeda 
r? @notriddle

ojeda commented:
That was quick, thanks a lot!

@rustbot label A-rust-for-linux

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12419215212/job/34674070591) [(plain)](https://github.com/rust-lang/rust/commit/ff979fdb4920038a5d7bc8d204da1e3495929fcb/checks/34674070591/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#13 3.256 Building wheels for collected packages: reuse
#13 3.257   Building wheel for reuse (pyproject.toml): started
#13 3.508   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 3.509   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132720 sha256=026f3bb0f1aa8090b861fd0a0939cb1a782396d84c8aab7875096557d637a0f6
#13 3.509   Stored in directory: /tmp/pip-ephem-wheel-cache-7wch0di6/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#13 3.512 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#13 3.913 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#13 3.914 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 4.449 Collecting virtualenv
#13 4.449 Collecting virtualenv
#13 4.489   Downloading virtualenv-20.28.0-py3-none-any.whl (4.3 MB)
#13 4.601      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.3/4.3 MB 39.0 MB/s eta 0:00:00
#13 4.642 Collecting distlib<1,>=0.3.7
#13 4.647   Downloading distlib-0.3.9-py2.py3-none-any.whl (468 kB)
#13 4.656      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 72.0 MB/s eta 0:00:00
#13 4.691 Collecting filelock<4,>=3.12.2
#13 4.695   Downloading filelock-3.16.1-py3-none-any.whl (16 kB)
#13 4.729 Collecting platformdirs<5,>=3.9.1
#13 4.734   Downloading platformdirs-4.3.6-py3-none-any.whl (18 kB)
#13 4.816 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 5.003 Successfully installed distlib-0.3.9 filelock-3.16.1 platformdirs-4.3.6 virtualenv-20.28.0
#13 DONE 5.1s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      108480 kB
DirectMap2M:     6182912 kB
DirectMap1G:    12582912 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
WARN: currently no CI rustc builds have rustc debug assertions enabled. Please either set `rust.debug-assertions` to `false` if you want to use download CI rustc or set `rust.download-rustc` to `false`.
downloading https://static.rust-lang.org/dist/2024-11-27/rustfmt-nightly-x86_64-unknown-linux-gnu.tar.xz
---
fmt check
fmt: checked 5733 files
tidy check
tidy: Skipping binary file check, read-only filesystem
tidy error: /checkout/src/doc/rustdoc/src/unstable-features.md: missing trailing newline
##[error]tidy error: /checkout/tests/rustdoc-ui/extract-doctests.rs:7: unexplained "```ignore" doctest; try one:
* make the test actually pass, by adding necessary imports and declarations, or
* make the test actually pass, by adding necessary imports and declarations, or
* use "```text", if the code is not Rust code, or
* use "```compile_fail,Ennnn", if the code is expected to fail at compile time, or
* use "```should_panic", if the code is expected to fail at run time, or
* use "```no_run", if the code should type-check but not necessary linkable/runnable, or
* explain it like "```ignore (cannot-test-this-because-xxxx)", if the annotation cannot be avoided.

removing old virtual environment
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (24.3.1)
All checks passed!
checking C++ file formatting
some tidy checks failed
some tidy checks failed
Command has failed. Rerun with -v to see more details.
  local time: Thu Dec 19 19:19:47 UTC 2024
  network time: Thu, 19 Dec 2024 19:19:48 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12419395182/job/34674632881) [(plain)](https://github.com/rust-lang/rust/commit/0b006d4514a221ea63716c5c013273596d8b8848/checks/34674632881/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#13 2.780 Building wheels for collected packages: reuse
#13 2.781   Building wheel for reuse (pyproject.toml): started
#13 3.030   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 3.030   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132720 sha256=026f3bb0f1aa8090b861fd0a0939cb1a782396d84c8aab7875096557d637a0f6
#13 3.031   Stored in directory: /tmp/pip-ephem-wheel-cache-u54eoo2u/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#13 3.034 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#13 3.434 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#13 3.434 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 3.959 Collecting virtualenv
#13 3.959 Collecting virtualenv
#13 3.995   Downloading virtualenv-20.28.0-py3-none-any.whl (4.3 MB)
#13 4.061      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.3/4.3 MB 67.6 MB/s eta 0:00:00
#13 4.115 Collecting platformdirs<5,>=3.9.1
#13 4.118   Downloading platformdirs-4.3.6-py3-none-any.whl (18 kB)
#13 4.153 Collecting filelock<4,>=3.12.2
#13 4.175 Collecting distlib<1,>=0.3.7
#13 4.178   Downloading distlib-0.3.9-py2.py3-none-any.whl (468 kB)
#13 4.186      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 79.3 MB/s eta 0:00:00
#13 4.186      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 79.3 MB/s eta 0:00:00
#13 4.267 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 4.455 Successfully installed distlib-0.3.9 filelock-3.16.1 platformdirs-4.3.6 virtualenv-20.28.0
#13 DONE 4.5s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      126912 kB
DirectMap2M:     7213056 kB
DirectMap1G:    11534336 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
WARN: currently no CI rustc builds have rustc debug assertions enabled. Please either set `rust.debug-assertions` to `false` if you want to use download CI rustc or set `rust.download-rustc` to `false`.
downloading https://static.rust-lang.org/dist/2024-11-27/rustfmt-nightly-x86_64-unknown-linux-gnu.tar.xz
---
fmt check
fmt: checked 5733 files
tidy check
tidy: Skipping binary file check, read-only filesystem
##[error]tidy error: /checkout/tests/rustdoc-ui/extract-doctests.rs:7: unexplained "```ignore" doctest; try one:
* make the test actually pass, by adding necessary imports and declarations, or
* make the test actually pass, by adding necessary imports and declarations, or
* use "```text", if the code is not Rust code, or
* use "```compile_fail,Ennnn", if the code is expected to fail at compile time, or
* use "```should_panic", if the code is expected to fail at run time, or
* use "```no_run", if the code should type-check but not necessary linkable/runnable, or
* explain it like "```ignore (cannot-test-this-because-xxxx)", if the annotation cannot be avoided.

removing old virtual environment
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (24.3.1)
All checks passed!
checking C++ file formatting
some tidy checks failed
some tidy checks failed
Command has failed. Rerun with -v to see more details.
  local time: Thu Dec 19 19:33:13 UTC 2024
  network time: Thu, 19 Dec 2024 19:33:13 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12430583871/job/34706311535) [(plain)](https://github.com/rust-lang/rust/commit/486acda9cc005a64107507056b9e9dfb14853745/checks/34706311535/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#21 exporting to docker image format
#21 sending tarball 27.4s done
#21 DONE 39.7s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
  Downloaded boml v0.3.1
   Compiling boml v0.3.1
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
    Finished `release` profile [optimized] target(s) in 3.97s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-system-gcc --use-backend gcc --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --mini-tests --std-tests`
Using system GCC
[BUILD] example
[AOT] mini_core_hello_world
/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc/mini_core_hello_world
abc
---
---- [ui] tests/rustdoc-ui/extract-doctests.rs stdout ----
Saved the actual stdout to "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc-ui/extract-doctests/extract-doctests.stdout"
diff of stdout:

- [{"filename":"tests/rustdoc-ui/extract-doctests.rs","line":7,"langstr":{"original":"ignore (checking attributes)","should_panic":false,"no_run":false,"ignore":"All","rust":true,"test_harness":false,"compile_fail":false,"standalone_crate":false,"error_codes":[],"edition":null,"added_classes":[],"unknown":[]},"text":"let x = 12;\nlet y = 14;"},{"filename":"tests/rustdoc-ui/extract-doctests.rs","line":12,"langstr":{"original":"edition2018,compile_fail","should_panic":false,"no_run":true,"ignore":"None","rust":true,"test_harness":false,"compile_fail":true,"standalone_crate":false,"error_codes":[],"edition":"2018","added_classes":[],"unknown":[]},"text":"let"}]
+ [{"filename":"$DIR/extract-doctests.rs","line":7,"langstr":{"original":"ignore (checking attributes)","should_panic":false,"no_run":false,"ignore":"All","rust":true,"test_harness":false,"compile_fail":false,"standalone_crate":false,"error_codes":[],"edition":null,"added_classes":[],"unknown":[]},"text":"let x = 12;\nlet y = 14;"},{"filename":"$DIR/extract-doctests.rs","line":12,"langstr":{"original":"edition2018,compile_fail","should_panic":false,"no_run":true,"ignore":"None","rust":true,"test_harness":false,"compile_fail":true,"standalone_crate":false,"error_codes":[],"edition":"2018","added_classes":[],"unknown":[]},"text":"let"}]
Note: some mismatched output was normalized before being compared
Note: some mismatched output was normalized before being compared
- [{"filename":"/checkout/tests/rustdoc-ui/extract-doctests.rs","line":7,"langstr":{"original":"ignore (checking attributes)","should_panic":false,"no_run":false,"ignore":"All","rust":true,"test_harness":false,"compile_fail":false,"standalone_crate":false,"error_codes":[],"edition":null,"added_classes":[],"unknown":[]},"text":"let x = 12;\nlet y = 14;"},{"filename":"/checkout/tests/rustdoc-ui/extract-doctests.rs","line":12,"langstr":{"original":"edition2018,compile_fail","should_panic":false,"no_run":true,"ignore":"None","rust":true,"test_harness":false,"compile_fail":true,"standalone_crate":false,"error_codes":[],"edition":"2018","added_classes":[],"unknown":[]},"text":"let"}]
+ [{"filename":"$DIR/extract-doctests.rs","line":7,"langstr":{"original":"ignore (checking attributes)","should_panic":false,"no_run":false,"ignore":"All","rust":true,"test_harness":false,"compile_fail":false,"standalone_crate":false,"error_codes":[],"edition":null,"added_classes":[],"unknown":[]},"text":"let x = 12;\nlet y = 14;"},{"filename":"$DIR/extract-doctests.rs","line":12,"langstr":{"original":"edition2018,compile_fail","should_panic":false,"no_run":true,"ignore":"None","rust":true,"test_harness":false,"compile_fail":true,"standalone_crate":false,"error_codes":[],"edition":"2018","added_classes":[],"unknown":[]},"text":"let"}]

The actual stdout differed from the expected stdout.
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args extract-doctests.rs`
To only update this specific test, also pass `--test-args extract-doctests.rs`

error: 1 errors occurred comparing output.
status: exit status: 0
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustdoc" "/checkout/tests/rustdoc-ui/extract-doctests.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc-ui/extract-doctests" "-A" "internal_features" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-Z" "unstable-options" "--extract-doctests"
--- stdout -------------------------------
[{"filename":"/checkout/tests/rustdoc-ui/extract-doctests.rs","line":7,"langstr":{"original":"ignore (checking attributes)","should_panic":false,"no_run":false,"ignore":"All","rust":true,"test_harness":false,"compile_fail":false,"standalone_crate":false,"error_codes":[],"edition":null,"added_classes":[],"unknown":[]},"text":"let x = 12;\nlet y = 14;"},{"filename":"/checkout/tests/rustdoc-ui/extract-doctests.rs","line":12,"langstr":{"original":"edition2018,compile_fail","should_panic":false,"no_run":true,"ignore":"None","rust":true,"test_harness":false,"compile_fail":true,"standalone_crate":false,"error_codes":[],"edition":"2018","added_classes":[],"unknown":[]},"text":"let"}]
stderr: none



```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12432557402/job/34712232541) [(plain)](https://github.com/rust-lang/rust/commit/67a0fef49ed8d741d60bc6a5ede880fdca85d9e9/checks/34712232541/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#21 exporting to docker image format
#21 sending tarball 27.6s done
#21 DONE 33.9s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
  Downloaded boml v0.3.1
   Compiling boml v0.3.1
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
    Finished `release` profile [optimized] target(s) in 3.96s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-system-gcc --use-backend gcc --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --mini-tests --std-tests`
Using system GCC
[BUILD] example
[AOT] mini_core_hello_world
/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc/mini_core_hello_world
abc
---
---- [run-make] tests/run-make/rustdoc-default-output stdout ----

error: rmake recipe failed to complete
status: exit status: 101
command: cd "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/rustdoc-default-output/rmake_out" && env -u RUSTFLAGS AR="ar" CARGO="/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools-bin/cargo" CC="cc" CC_DEFAULT_FLAGS="-ffunction-sections -fdata-sections -fPIC -m64" CXX="c++" CXX_DEFAULT_FLAGS="-ffunction-sections -fdata-sections -fPIC -m64" HOST_RPATH_DIR="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" LD_LIBRARY_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-bootstrap-tools/x86_64-unknown-linux-gnu/release/deps:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools-bin:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" LD_LIB_PATH_ENVVAR="LD_LIBRARY_PATH" LLVM_BIN_DIR="/usr/lib/llvm-18/bin" LLVM_COMPONENTS="aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets amdgpu amdgpuasmparser amdgpucodegen amdgpudesc amdgpudisassembler amdgpuinfo amdgputargetmca amdgpuutils analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo cfguard codegen codegentypes core coroutines coverage debuginfobtf debuginfocodeview debuginfodwarf debuginfogsym debuginfologicalview debuginfomsf debuginfopdb demangle dlltooldriver dwarflinker dwarflinkerclassic dwarflinkerparallel dwp engine executionengine extensions filecheck frontenddriver frontendhlsl frontendoffloading frontendopenacc frontendopenmp fuzzercli fuzzmutate globalisel hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo hipstdpar instcombine instrumentation interfacestub interpreter ipo irprinter irreader jitlink lanai lanaiasmparser lanaicodegen lanaidesc lanaidisassembler lanaiinfo libdriver lineeditor linker loongarch loongarchasmparser loongarchcodegen loongarchdesc loongarchdisassembler loongarchinfo lto m68k m68kasmparser m68kcodegen m68kdesc m68kdisassembler m68kinfo mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts objcopy object objectyaml option orcdebugging orcjit orcshared orctargetprocess passes perfjitevents powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo riscvtargetmca runtimedyld scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target targetparser textapi textapibinaryreader transformutils ve veasmparser vecodegen vectorize vedesc vedisassembler veinfo webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo webassemblyutils windowsdriver windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info x86targetmca xcore xcorecodegen xcoredesc xcoredisassembler xcoreinfo xray xtensa xtensaasmparser xtensacodegen xtensadesc xtensadisassembler xtensainfo" LLVM_FILECHECK="/usr/lib/llvm-18/bin/FileCheck" NODE="/usr/bin/node" PYTHON="/usr/bin/python3" RUSTC="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" RUSTDOC="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustdoc" SOURCE_ROOT="/checkout" TARGET="x86_64-unknown-linux-gnu" TARGET_RPATH_DIR="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" TARGET_RPATH_ENV="/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/rustdoc-default-output/rmake_out:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-bootstrap-tools/x86_64-unknown-linux-gnu/release/deps:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/rustdoc-default-output/rmake"
--- stderr -------------------------------
thread 'main' panicked at /checkout/tests/run-make/rustdoc-default-output/rmake.rs:15:10:
test failed: `output-default.stdout` is different from `actual`

---
                         Disable HTML source code pages generation
+        --extract-doctests 
+                        Output doctests in JSON format
 
     @path               Read newline separated options from `path`

note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
------------------------------------------

```

</details>

notriddle commented on src/doc/rustdoc/src/unstable-features.md:
````diff
@@ -664,3 +664,22 @@ Similar to cargo `build.rustc-wrapper` option, this flag takes a `rustc` wrapper
+Using this flag looks like this:
+
+```bash
+$ rustdoc -Zunstable-options --extract-doctests src/lib.rs
````
Why is this a CLI option and not an output format?

GuillaumeGomez replied:
Because I didn't think about it, it's SO MUCH better!

notriddle commented on src/doc/rustdoc/src/unstable-features.md:
```diff
@@ -664,3 +664,22 @@ Similar to cargo `build.rustc-wrapper` option, this flag takes a `rustc` wrapper
+When this flag is used, it outputs the doctests original source code alongside information
+such as:
+
+ * File where they are located.
+ * Line where they are located.
+ * Codeblock attributes (more information about this [here](./write-documentation/documentation-tests.html#attributes)).
+
+The output format is JSON.
```
That's not enough documentation, even for an unstable feature. You need to say what the actual keys are, and what they do (particularly the subkeys of `langstr`, which are not self-explanatory). Please include a prettified example JSON document with all of the format features used.

GuillaumeGomez replied:
I'm not too sure if we want everything to be documented as docblock attributes list might get longer. Well, I'll list the current one and we can think about it later.

aDotInTheVoid commented:
I don't think the approach taken here (or manually implementing `serde::Serialize` for internal structs) is a good one.

It means some of the internal structs are actually part of the public API, but the exact relation is unclear in source code.

I think the right way to structure this is to do what `--output-format json` does and keep the type definitions for (de)serialization very separate from the all rustc_*/rustdoc internal types, and have them in there own module/crate. This achieves:

- It's very clear when a PR adjusts the schema (whereas at the moment, that's not super apparent)
- We can just use `#[derive(Serialize, Deserialize)]` on the module (rather than having to implement manually)
- We can republish the module onto crates.io, so it's easy for (non-kernel) rust users (eg nextest, etc)
- We can the modules api-docs to describe the schema.

Some other things:
- There should be an option (probably the default) to output to a file.
- There should be a `FORMAT_VERSION` field at the root of the output. This was super critical for rustdoc-json to be able to evolve without breaking users in a really terrible way.

aDotInTheVoid commented on tests/rustdoc-ui/extract-doctests.rs:
```diff
@@ -0,0 +1,15 @@
+//@ compile-flags:-Z unstable-options --extract-doctests
+//@ normalize-stdout-test: "tests/rustdoc-ui" -> "$$DIR"
+//@ check-pass
+
```
We probably want a test for how this interacts with:

- Implicit/explicit `fn main()`
- Hidding lines with `/// # use some::path;`

GuillaumeGomez replied:
True although currently it's the code as defined in the documentation and no changes on rustdoc side (which I think matches better what rust for linux wants).

GuillaumeGomez replied:
Maybe we should provide another field with "rustdoc computed code" where hidden lines and `main` function wrapping are added by rustdoc. Actually that sounds like a very good idea.

ojeda replied:
> True although currently it's the code as defined in the documentation and no changes on rustdoc side (which I think matches better what rust for linux wants).

Currently we use both hidden lines and the `?` support, i.e. the `# Ok::<..., ...>(...)` syntax (which implies the `fn main()`, from what I understand).

Some of that post-processing may be easy to do by users, like the hidden lines I assume, but if you walk the source or IR or similar to figure out details (like the crate attributes that you move out of `main()`), then it may be harder for end users to replicate that properly without a hack. Perhaps it may be possible to export what `rustdoc` figured about the test, so that users can replicate the post-processing on their side, but customized to their environment.

Having the `rustdoc` computed code sounds fine since it may be enough for some use cases, but e.g. we currently need to convert on the fly the `.unwrap()` in the `?`-using tests into a custom `assert!`.

ojeda replied:
So, for instance, if `rustdoc` tells us "this test uses `?`", "these are the crate attributes I would have moved", etc., then users may be able to easily and reliably construct their own wrappers and e.g. do a check instead of an `.unwrap()`.

Of course, for things like crate attributes, it may be best to remove them so that the end user can re-add them where needed, so it wouldn't be the completely "unaltered" source code. But it could be an "adapted" version. Hidden lines should probably be normal lines in that adapted version.

Perhaps it makes sense to provide all those versions, i.e. the completely unaltered one for those that may want to do something complex or to render the text somewhere, the "adapted" version for customized test environments and the `rustdoc` computed code for those that can use directly that. In the kernel we would use the "adapted" one, only, I think.

GuillaumeGomez commented:
> I don't think the approach taken here (or manually implementing `serde::Serialize` for internal structs) is a good one.
> 
> It means some of the internal structs are actually part of the public API, but the exact relation is unclear in source code.
> 
> I think the right way to structure this is to do what `--output-format json` does and keep the type definitions for (de)serialization very separate from the all rustc_*/rustdoc internal types, and have them in there own module/crate. This achieves:
> 
>     * It's very clear when a PR adjusts the schema (whereas at the moment, that's not super apparent)
> 
>     * We can just use `#[derive(Serialize, Deserialize)]` on the module (rather than having to implement manually)
> 
>     * We can republish the module onto crates.io, so it's easy for (non-kernel) rust users (eg nextest, etc)
> 
>     * We can the modules api-docs to describe the schema.

I think you're right, I took the easiest way which has the advantage to always keep the output up-to-date since it uses the same types for the code generation. I'll split it out.

> Some other things:
> 
>     * There should be an option (probably the default) to output to a file.

I'm not convinced that adding an option to output to a file is needed. I have these two cases in mind:
 1. From a shell you redirect stdout.
 2. From a program you can just get the stdout once again and then do what you need with it.

Do you think another case I didn't think about maybe?

>     * There should be a `FORMAT_VERSION` field at the root of the output. This was super critical for rustdoc-json to be able to evolve without breaking users in a really terrible way.

Good catch, gonna add it, thanks!

aDotInTheVoid commented:
> I took the easiest way which has the advantage to always keep the output up-to-date since it uses the same types for the code generation.

I think the nicest way to get this is to unpack the “internal” types exhaustively in the internal->public conversion. This way you’re forced to think about the public repr when internals change, but they’re still seperate and easy to track.

> Do you think another case I didn't think about maybe?

I think the big one is build-systems, which are much more used to chaining commands together via files than stdout.

bors commented:
:umbrella: The latest upstream changes (presumably #134590) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

GuillaumeGomez committed:
Add new output-format

GuillaumeGomez committed:
Add ui test for new rustdoc `--output-format=doctest` option

GuillaumeGomez committed:
Move extracted doctest code and types into its own file

GuillaumeGomez committed:
Mention the tracking issue of `--output-format=doctest`

GuillaumeGomez committed:
Add documentation for `--output-format=doctest`

GuillaumeGomez commented:
Big update:

I separated output type from rustdoc's type as suggested by @aDotInTheVoid as it will make it much easier for us to maintain.

I added a `format_version` field.

I added the modified rustdoc code as well (in a `doctest_code` field).

I changed the option to `--output-format=doctest`.

I greatly improved the book description.

Hopefully I didn't miss anything. :)

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12994310354/job/36238526483) [(plain)](https://github.com/rust-lang/rust/commit/276a3a1701da156cd5f24e5e82b5d54db6d86b60/checks/36238526483/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
info: removing rustup binaries
info: rustup is uninstalled
##[group]Image checksum input
mingw-check-tidy
# We use the ghcr base image because ghcr doesn't have a rate limit
# and the mingw-check-tidy job doesn't cache docker images in CI.
FROM ghcr.io/rust-lang/ubuntu:22.04
ARG DEBIAN_FRONTEND=noninteractive
RUN apt-get update && apt-get install -y --no-install-recommends \
  g++ \
  make \
---

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#12 4.658      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.3/4.3 MB 51.5 MB/s eta 0:00:00
#12 4.703 Collecting distlib<1,>=0.3.7
#12 4.714   Downloading distlib-0.3.9-py2.py3-none-any.whl (468 kB)
#12 4.722      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 78.2 MB/s eta 0:00:00
#12 4.765 Collecting platformdirs<5,>=3.9.1
#12 4.776   Downloading platformdirs-4.3.6-py3-none-any.whl (18 kB)
#12 4.820 Collecting filelock<4,>=3.12.2
#12 4.831   Downloading filelock-3.17.0-py3-none-any.whl (16 kB)
#12 4.915 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#12 5.101 Successfully installed distlib-0.3.9 filelock-3.17.0 platformdirs-4.3.6 virtualenv-20.29.1
#12 DONE 5.2s

#13 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#13 DONE 0.0s
---
DirectMap4k:      122816 kB
DirectMap2M:     7217152 kB
DirectMap1G:    11534336 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
    Finished `dev` profile [unoptimized] target(s) in 0.05s
##[endgroup]
WARN: currently no CI rustc builds have rustc debug assertions enabled. Please either set `rust.debug-assertions` to `false` if you want to use download CI rustc or set `rust.download-rustc` to `false`.
downloading https://static.rust-lang.org/dist/2025-01-08/rustfmt-nightly-x86_64-unknown-linux-gnu.tar.xz
---
fmt check
Diff in /checkout/src/librustdoc/doctest/extracted.rs:1:
+use serde::Serialize;
+
+use super::{DocTestBuilder, ScrapedDocTest};
 use crate::config::Options as RustdocOptions;
 use crate::html::markdown;
-use super::{DocTestBuilder, ScrapedDocTest};
-use serde::Serialize;
 const FORMAT_VERSION: u32 = 1;
 
Diff in /checkout/src/librustdoc/doctest/extracted.rs:14:
 
---
-        }
+        Self { format_version: FORMAT_VERSION, doctests: Vec::new() }
     }
 
-    pub(crate) fn add_test(&mut self, scraped_test: ScrapedDocTest, opts: &super::GlobalTestOptions, options: &RustdocOptions) {
+        &mut self,
+        scraped_test: ScrapedDocTest,
+        scraped_test: ScrapedDocTest,
+        opts: &super::GlobalTestOptions,
+        options: &RustdocOptions,
+    ) {
         let edition = scraped_test.edition(&options);
-        let ScrapedDocTest {
-            filename,
-            line,
-            langstr,
```

</details>

GuillaumeGomez commented:
And fixed `fmt`.

aDotInTheVoid commented on src/librustdoc/config.rs:
```diff
@@ -446,12 +448,22 @@ impl Options {
-        if !matches!(matches.opt_str("output-format").as_deref(), None | Some("html"))
+        if let Some(format) = matches.opt_str("output-format").as_deref()
+            && format != "html"
             && !matches.opt_present("show-coverage")
```
What happens if I run `rustdoc --output-format=doctest --show-coverage`

GuillaumeGomez replied:
Good point! Gonna also add a UI test for this case.

aDotInTheVoid commented on src/doc/rustdoc/src/unstable-features.md:
````diff
@@ -542,6 +544,67 @@ It can also be used with `--show-coverage`. Take a look at its
+pub trait Trait {}
+```
+
+The generated output will look like this:
````
This should document that the output goes to `stdout`

aDotInTheVoid commented on src/librustdoc/doctest.rs:
```diff
@@ -226,21 +228,42 @@ pub(crate) fn run(dcx: DiagCtxtHandle<'_>, input: Input, options: RustdocOptions
 
         collector
-    });
+    }) {
```
This control-flow is hard to follow. Given that the only source of `Ok(None)` and `Err` are in the `exctract_doctests` branch, the early return should be moved up there (so when reading them there it's clear that they end the function)

GuillaumeGomez replied:
We can't because it'd leave the closure, not the whole `run` function. Hence why I had to do it in two steps. I can improve the code by storing the `run_compiler` returned value in a variable though.

aDotInTheVoid commented on src/librustdoc/doctest/extracted.rs:
```diff
@@ -0,0 +1,132 @@
+use serde::Serialize;
```
There should be a doc-comment here explaining what feature this is supporting, and that `FORMAT_VERSION` needs to be bumped when the types are changed.

GuillaumeGomez replied:
Very good point!

aDotInTheVoid commented on src/doc/rustdoc/src/unstable-features.md:
```diff
@@ -542,6 +544,67 @@ It can also be used with `--show-coverage`. Take a look at its
+   * `line` is the line where the doctest starts (so where the \`\`\` is located in the current code).
+   * `doctest_attributes` contains computed information about the attributes used on the doctests. For more information about doctest attributes, take a look [here](write-documentation/documentation-tests.html#attributes).
+   * `original_code` is the code as written in the source code before rustdoc modifies it.
+   * `doctest_code` is the code modified by rustdoc that will be run. If there is a syntax error, this field will not be present.
```
This isn't right: From the doctest, we output `"doctest_code":"#![allow(unused)]\nfn main() {\nlet\n}"`

GuillaumeGomez replied:
I don't really know how to word it. In this case the AST isn't invalid, it's incomplete (which is not the same for the compiler).

aDotInTheVoid commented on tests/rustdoc-ui/extract-doctests.stdout:
```diff
@@ -0,0 +1 @@
+{"format_version":1,"doctests":[{"file":"$DIR/extract-doctests.rs","line":8,"doctest_attributes":{"original":"ignore (checking attributes)","should_panic":false,"no_run":false,"ignore":"All","rust":true,"test_harness":false,"compile_fail":false,"standalone_crate":false,"error_codes":[],"edition":null,"added_css_classes":[],"unknown":[]},"original_code":"let x = 12;\nlet y = 14;","doctest_code":"# ...
```
I still don't love this as a way of testing, but it's probably fine for now. Eventually, `run-make` is probably nicer.

GuillaumeGomez replied:
I wanted to re-use the `rustdoc-json` testsuite originally but it'd require some extra changes. So for now I have put it here and I plan to send a follow-up where I add an option allowing `rustdoc-json` tests to use `--output-format=doctest` instead.

aDotInTheVoid commented on src/librustdoc/lib.rs:
```diff
@@ -814,7 +814,12 @@ fn main_args(early_dcx: &mut EarlyDiagCtxt, at_args: &[String]) {
+        config::markdown_input(&input),
+    ) {
         (true, Some(_)) => return wrap_return(dcx, doctest::test_markdown(&input, options)),
         (true, None) => return doctest::run(dcx, input, options),
```
`doctest::run` should be renamed, because now it either runs *or* just collects and prints the doctests.

GuillaumeGomez replied:
We run the `doctest` pass. Same for outputs named `run_renderer`. I don't mind changing it though, how would you rename it?

aDotInTheVoid reviewed (commented):
This is looking in alot better shape, thanks.

I think it's fine to land on nightly while the public API uses the internal types, I'd like to move away from that eventually, but that can be done in a follow-up PR. Same with test improvements.


@notriddle Are you happy for me to r+ this after I'm happy with with it, or should I also make sure you're good with the current version before merging?

GuillaumeGomez committed:
Improve code and add missing docs for new `doctest::extracted` module

GuillaumeGomez committed:
Improve check for `--output-format` combinations and add ui regression test

GuillaumeGomez commented:
Applied suggestions.

notriddle commented:
> Are you happy for me to r+ this after I'm happy with with it, or should I also make sure you're good with the current version before merging?

r=me

GuillaumeGomez commented:
Then let's go. Thanks to both of you!

@bors r=notriddle,aDotInTheVoid rollup

bors commented:
:pushpin: Commit b7951380ca9e8eddcd67bf947b592d10e21bdea5 has been approved by `notriddle,aDotInTheVoid`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=notriddle,aDotInTheVoid b7951380ca9e8eddcd67bf947b592d10e21bdea5 -->
<!-- homu: {"type":"Approved","sha":"b7951380ca9e8eddcd67bf947b592d10e21bdea5","approver":"notriddle,aDotInTheVoid","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
